mesh/node.c

```diff
@@ -1696,25 +1797,61 @@ static bool node_application_keys_getter(struct l_dbus *dbus,
 				struct l_dbus_message_builder *builder,
 				void *user_data)
 {
-	const uint8_t app_keys[2][16] = {
-		{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
-		{0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1}
-	};
-	int i;
+	struct mesh_node *node;
+	struct l_queue *app_keys_queue;
+	unsigned int app_keys_count;
+	const char *path;
+	uint8_t uuid[KEY_LEN];
+	uint8_t **app_keys;
+	unsigned int i;
+	bool status = true;
 
-	//TODO - create api to get application keys from node
+	path = l_dbus_message_get_path(message);
+	if (!get_uuid_from_path(path, uuid))
+		goto failed;
 
-	if (!l_dbus_message_builder_enter_array(builder, "ay"))
-		return false;
+	node = l_queue_find(nodes, match_node_uuid, uuid);
+	if (!node)
+		goto failed;
 
-	for (i = 0; i < 2; i++)
-		if (!dbus_append_byte_array(builder, app_keys[i], 16))
-			return false;
+	if (!node->net)
+		goto failed;
 
-	if (!l_dbus_message_builder_leave_array(builder))
-		return false;
+	app_keys_queue = mesh_net_get_app_keys(node->net);
+	if (!app_keys_queue)
+		goto failed;
 
-	return true;
+	app_keys_count = l_queue_length(app_keys_queue);
+	if (app_keys_count == 0)
+		goto failed;
+
+	app_keys = (uint8_t **) l_new(uint8_t*, app_keys_count);
```

Comment:
As far as I remember, app key indexes don't need to be consecutive, so we should rather return a dict, instead or re-indexing keys.

Also, take a look at `l_queue_foreach`